src/net/ossl.cc

```diff
@@ -1657,9 +1602,187 @@ class session : public enable_shared_from_this<session>, public session_impl {
     session_type _type;
     bool _eof = false;
     bool _shutdown = false;
+
+    friend int bio_write_ex(BIO* b, const char * data, size_t dlen, size_t * written);
+    friend int bio_read_ex(BIO* b, char * data, size_t dlen, size_t *readbytes);
+    friend long bio_ctrl(BIO * b, int ctrl, long num, void * data);
 };
+
+
+tls::session* unwrap_bio_ptr(void * ptr) {
+    return static_cast<tls::session*>(ptr);
+}
+
+tls::session* unwrap_bio_ptr(BIO * b) {
+    return unwrap_bio_ptr(BIO_get_data(b));
+}
+
+/// The 'ioctl' for BIO
+long bio_ctrl(BIO * b, int ctrl, long num, void * data) {
+    if (BIO_get_init(b) <= 0 && ctrl != BIO_C_SET_POINTER) {
+        return 0;
+    }
+
+    auto session = unwrap_bio_ptr(b);
+
+    switch(ctrl) {
+    case BIO_C_SET_POINTER:
+        if (BIO_get_init(b) <= 0) {
+            BIO_set_data(b, data);
+            BIO_set_init(b, 1);
+            return 1;
+        } else {
+            return 0;
+        }
+    case BIO_CTRL_GET_CLOSE:
+        return BIO_get_shutdown(b);
+    case BIO_CTRL_SET_CLOSE:
+        BIO_set_shutdown(b, static_cast<int>(num));
+        break;
+
+    case BIO_CTRL_DUP:
+    case BIO_CTRL_FLUSH:
+        return 1;
+    case BIO_CTRL_EOF:
+        return BIO_test_flags(b, BIO_FLAGS_IN_EOF) != 0;
+    case BIO_CTRL_PENDING:
+        return static_cast<long>(session->_input.size());
+    case BIO_CTRL_WPENDING:
+        return session->_output_pending.available() ? 0 : 1;
+    default:
+        return 0;
+    }
+
+    return 0;
+}
+
+/// This is called when the BIO is created
+///
+/// It is important for this to be set, even if it doesn't do anything
+/// because the BIO init flag won't get automatically set to '1'.
+int bio_create(BIO*) {
+    return 1;
+}
+
+/// Handles writes to the BIO
+///
+/// This function will attempt to call _out.put() and store the future in
+/// _output_pending.  If _output_pending has not yet resolved, return '0'
+/// and set the retry write flag.
+int bio_write_ex(BIO* b, const char * data, size_t dlen, size_t * written) {
+    auto session = unwrap_bio_ptr(b);
+    BIO_clear_retry_flags(b);
+
+    if (!session->_output_pending.available()) {
+        BIO_set_retry_write(b);
+        return 0;
+    }
+
+    try {
+        size_t n;
+
+        if (!session->_output_pending.failed()) {
+            scattered_message<char> msg;
+            msg.append(std::string_view(data, dlen));
+            n = msg.size();
+            session->_output_pending = session->_out.put(std::move(msg).release());
+        }
+
+        if (session->_output_pending.failed()) {
+            std::rethrow_exception(session->_output_pending.get_exception());
+        }
```

Comment:
The comment of this function says that `If _output_pending has not yet resolved, return '0' and set the retry write flag.`

Do we need to call `BIO_set_retry_write(b);` here if `session->_output_pending` has not resolved yet?

Reply:
In this situation `_output_pending` has resolved but in a failure state, so we will throw an exception here (which is caught below) and `_output_pending` is set to be an exceptional future.  The function returns 0 and I add `ERR_LIB_SYS` to the error stack.  Internall, if OpenSSL sees that an `ERR_LIB_SYS` is at the top of the stack, `SSL_get_error` will return `SSL_ERROR_SYSCALL` which I then handle.

Reply:
Right, what I meant is for the case when we don't enter this if branch and `session->_output_pending` is not yet available. Do we need to set `BIO_set_retry_write(b);` in an `else` branch for this `if`?

Reply:
That happens at the very beginning of the function.  If upon the call to `bio_write_ex`, the `_output_pending` future is not yet resolved, then I return 0 and set the retry flag.  I don't expect `_output_pending` to resolve immediately after I call `_out.put`

Reply:
So the retry flag is there for backpressure when there is an outstanding write, whereas here we've successfully handled the write so there's no need to set a retry flag.